### ISSUE TITLE
Blurred player textures when bi/trilinear filtering is on

### ISSUE THREAD

**Issue by lhofhansl** [minetest/minetest#6556]
The player textures are very satisfyingly blurred when any filtering is enabled.
I think it would be best to not filter the player textures... Or is there something else wrong?

(I was randomly looking at MC imagery and the player textures just looked sharper there, and I began to pay attention to them in MT.)

Looks like filtering is applied in `TextureSource::generateTextureFromMesh`. If I distinguish player (or mob...?) textures from node ones I could make a - perhaps configurable - change there.

**Comment by lhofhansl** [comment]
Found the spot: `GenericCAO::updateTextures`. Now the question is: What to do about it?
Personally I prefer players and all mobs not to have any filtering done on their textures, makes them look ... no sharp.
I imagine others feel differently...? Should I make this configurable?

**Comment by paramat** [comment]
Do you mean that MT player/mob textures seem more filtered than the world, or are they consistent?
I would need to see the difference, but my first thought is whether this might make players / mobs seem not a part of the world.

**Comment by lhofhansl** [comment]
I'll do a PR to try. Right now it looks like players are far more blurred than the rest of the world. As said for players/mobs I prefer no filtering at all.

**Comment by Fixer-007** [comment]
It is not just player, but everything else is also blurred in unpleasant way.

**Comment by raymoo** [comment]
The problem is that the filtering is done without taking into account the magnification of the textures. Texture in MT are generally supposed to look pixelated, so they need to be filtered "as if" they were magnified to their displayed size. This means areas in the middle of the pixel have a flat color as desired, while areas near pixel borders will be blurred. You still want to filter because you will get aliasing if you don't, especially when minifying.

If Irrlicht does not support filtering in this way, it may be an Irrlicht problem. But changing texture filtering for players to nearest neighbor will not really solve the problem.

**Comment by lhofhansl** [comment]
I'll do a PR to try. Right now it looks like players are far more blurred than the rest of the world. As said for players/mobs I prefer no filtering at all.

**Comment by lhofhansl** [comment]
To turn filtering off for meshes, this is all it takes (commenting the filtering out)... In case you wanted to test how it looks.

```
diff --git a/src/content_cao.cpp b/src/content_cao.cpp
index 7178deba..9793c0ec 100644
--- a/src/content_cao.cpp
+++ b/src/content_cao.cpp
@@ -1027,13 +1027,14 @@ void GenericCAO::updateTextures(std::string mod)
                                material.TextureLayer[0].Texture = texture;
                                material.setFlag(video::EMF_LIGHTING, false);
                                material.setFlag(video::EMF_BILINEAR_FILTER, false);
-
+                               /*
                                m_animated_meshnode->getMaterial(i)
                                                .setFlag(video::EMF_TRILINEAR_FILTER, use_trilinear_filter);
                                m_animated_meshnode->getMaterial(i)
                                                .setFlag(video::EMF_BILINEAR_FILTER, use_bilinear_filter);
                                m_animated_meshnode->getMaterial(i)
                                                .setFlag(video::EMF_ANISOTROPIC_FILTER, use_anisotropic_filter);
+                               */
                        }
                        for (u32 i = 0; i < m_prop.colors.size() &&
                        i < m_animated_meshnode->getMaterialCount(); ++i)
```

**Comment by raymoo** [comment]
How does it look with anisotropic filtering?

**Comment by lhofhansl** [comment]
@raymoo same blurryness. The problem with the player and some of the mob meshes is that their resolution is too low. It seems the filtering should happen after scaling them up with a much higher resolution.
I think that's what you are alluding to above.

I'll admit I know nothing about Irrlicht or how one would achieve this.

**Comment by raymoo** [comment]
@lhofhansl It doesn't have to actually scale them up, it can just change which texels are sampled for filtering. Like if you would normally sample the positions 1 to the left and 1 to the right of a texture coordinate, you would sample the positions 0.5 to the left and 0.5 to the right when the texture appears 2x as large as normal. I don't know how this would be achieved, but to me it seems like it would be something to set in Irrlicht.

**Comment by lhofhansl** [comment]
Found this discussion of the forum: [https://forum.minetest.net/viewtopic.php?pid=40417](https://forum.minetest.net/viewtopic.php?pid=40417)

**Comment by lhofhansl** [comment]
I was thinking along the line of the @VanessaE says there. When resolution of the texture is "high enough" then use the filter, otherwise don't. Would be nice to filter at a higher resolution than the texture, but I haven't found a way to do that, yet.

**Comment by Fixer-007** [comment]
>How does it look with anisotropic filtering?

If filtering is disabled - it looks sharp. I've always had "filtering off" and "mipmaps + aniso" "on"

**Comment by lhofhansl** [comment]
So anything to do here? I have a patch that puts a threshold of filtering for animated meshes to 64. So the world is still filtered, higher-res meshes are filtered, but player textures and other equally low-res meshes are not filtered.

I'm using high-res texture packs and those I'd very much like to be filtered, while the player just looks blurred.

Then again... Maybe it's not worth the effort.

**Comment by lhofhansl** [comment]
Like this:

```
diff --git a/src/content_cao.cpp b/src/content_cao.cpp
index 7178deba..8bc9ad88 100644
--- a/src/content_cao.cpp
+++ b/src/content_cao.cpp
@@ -1028,6 +1028,13 @@ void GenericCAO::updateTextures(std::string mod)
                                material.setFlag(video::EMF_LIGHTING, false);
                                material.setFlag(video::EMF_BILINEAR_FILTER, false);
 
+                               // don't filter low-res textures, makes them look blurry
+                               // player models have a res of 64
+                               const core::dimension2d<u32> &size = texture->getOriginalSize();
+                               const u32 res = std::min(size.Height, size.Width);
+                               use_trilinear_filter &= res > 64;
+                               use_bilinear_filter &= res > 64;
+
                                m_animated_meshnode->getMaterial(i)
                                                .setFlag(video::EMF_TRILINEAR_FILTER, use_trilinear_filter);
                                m_animated_meshnode->getMaterial(i)
```

**Comment by paramat** [comment]
I do think having players more blurry than the world is bad and should be addressed by either reducing the effect or removing it when necessary. So this seems worth doing.

**Comment by paramat** [comment]
![screenshot_20171027_033327](https://user-images.githubusercontent.com/3686677/32085534-aabcc7f6-bac7-11e7-8400-5f064bd42ac5.png)

Wow i had no idea it was this bad, the world seems barely affected.

**Comment by lhofhansl** [comment]
Thanks. OK. I'll do a PR.

**Comment by lhofhansl** [comment]
Fixed with #6562
